File: src/main/java/io/github/hds/pemu/processor/Processor.java
package io.github.hds.pemu.processor;

import io.github.hds.pemu.instructions.Instruction;
import io.github.hds.pemu.instructions.InstructionError;
import io.github.hds.pemu.instructions.InstructionHistory;
import io.github.hds.pemu.instructions.InstructionSet;
import io.github.hds.pemu.localization.Translation;
import io.github.hds.pemu.localization.TranslationManager;
import io.github.hds.pemu.memory.*;
import io.github.hds.pemu.memory.flags.*;
import io.github.hds.pemu.memory.registers.*;
import io.github.hds.pemu.utils.StringUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.awt.event.KeyEvent;

public final class Processor implements IProcessor {

    private boolean isRunning = false;

    // How many words the registers will occupy
    private final int REGISTERS_WORDS = 2;
    // How many words are reserved for the stack
    private final int RESERVED_STACK_ELEMENTS = 1;
    // How many words are reserved for flags
    private final int FLAGS_WORDS = 1;

    private final RegisterHolder<MemoryRegister> REGISTERS;
    private final FlagHolder<MemoryFlag> FLAGS;

    private final Memory MEMORY;
    private final Clock CLOCK;

    private final InstructionSet INSTRUCTIONSET;
    private final InstructionHistory HISTORY;

    private volatile char charPressed = '\0';
    private volatile int keyPressed = KeyEvent.VK_UNDEFINED;
    private long startTimestamp = 0;

    private volatile boolean isPaused = false;
    private volatile boolean stepping = false;

    public Processor(@NotNull ProcessorConfig config) {
        MEMORY = new Memory(
                config.getMemorySize(),
                Word.getClosestWord(config.getBits())
        );
        CLOCK = new Clock(config.getClockFrequency());

        INSTRUCTIONSET = config.getInstructionSet();
        HISTORY = new InstructionHistory();

        REGISTERS = new RegisterHolder<>(
                new MemoryRegister(getProgramAddress(), "Instruction Pointer", MEMORY, 0),
                new MemoryRegister(MEMORY.getSize() - 1, "Stack Pointer", MEMORY, 1)
        );

        FLAGS = new FlagHolder<>(
                new MemoryFlag(false, "Zero Flag" , MEMORY, REGISTERS_WORDS, 0),
                new MemoryFlag(false, "Carry Flag", MEMORY, REGISTERS_WORDS, 1)
        );
    }

    public static @NotNull DummyProcessor getDummyProcessor(@NotNull ProcessorConfig config) {
        return new DummyProcessor(
                config,
                new IDummyRegister[] {
                        new DummyMemoryRegister("Instruction Pointer"),
                        new DummyMemoryRegister("Stack Pointer")
                },
                new IDummyFlag[] {
                        new DummyMemoryFlag("Zero Flag"),
                        new DummyMemoryFlag("Carry Flag"),
                }
        );
    }

    @Override
    public @NotNull IFlag[] getFlags() {
        return FLAGS.toArray();
    }

    @Override
    public @NotNull IRegister[] getRegisters() {
        return REGISTERS.toArray();
    }

    @Override
    public @Nullable IFlag getFlag(@NotNull String shortName) {
        return FLAGS.getFlag(shortName);
    }

    @Override
    public @Nullable IRegister getRegister(@NotNull String shortName) {
        return REGISTERS.getRegister(shortName);
    }

    @Override
    public @NotNull IMemory getMemory() {
        return MEMORY;
    }

    @Override
    public @NotNull Clock getClock() {
        return CLOCK;
    }

    @Override
    public @NotNull InstructionSet getInstructionSet() {
        return INSTRUCTIONSET;
    }

    @Override
    public int getKeyPressed() {
        return keyPressed;
    }

    @Override
    public void setKeyPressed(int key) {
        keyPressed = key;
    }

    @Override
    public char getCharPressed() {
        return charPressed;
    }

    @Override
    public void setCharPressed(char ch) {
        charPressed = ch;
    }

    @Override
    public @NotNull String getInfo() {
        Translation currentTranslation = TranslationManager.getCurrentTranslation();
        return StringUtils.format(
                ("\t{0}:\t" + StringUtils.getEngNotation(CLOCK.getFrequency(), "Hz") + "\n" +
                 "\t{1}:\t" + MEMORY.getSize() + 'x' + MEMORY.getWord().TOTAL_BYTES + " Bytes\n" +
                 "\t{2}:\t" + INSTRUCTIONSET.getSize() + "\n"),
                currentTranslation.getOrDefault("messages.clock"),
                currentTranslation.getOrDefault("messages.memory"),
                currentTranslation.getOrDefault("messages.instructions")
        );
    }

    @Override
    public @Nullable InstructionHistory getInstructionHistory() {
        return HISTORY;
    }

    @Override
    public long getTimeRunning() {
        if (!isRunning) return -1;
        return System.currentTimeMillis() - startTimestamp;
    }

    @Override
    public @Nullable String loadProgram(int[] program) {
        if (program.length > MEMORY.getSize() - getReservedWords())
            return TranslationManager.getCurrentTranslation().getOrDefault("messages.processorOutOfMemory");

        MEMORY.setValuesAt(getProgramAddress(), program);
        return null;
    }

    @Override
    public int getProgramAddress() {
        return REGISTERS_WORDS + FLAGS_WORDS;
    }

    @Override
    public int getReservedWords() {
        return REGISTERS_WORDS + FLAGS_WORDS + RESERVED_STACK_ELEMENTS;
    }

    @Override
    public boolean isRunning() {
        return this.isRunning;
    }

    @Override
    public void run() {
        if (isRunning) return;

        startTimestamp = System.currentTimeMillis();
        isRunning = true;
        while (isRunning) {
            if (CLOCK.update() && (stepping || !isPaused)) {
                stepping = false;

                final MemoryRegister IP = REGISTERS.getRegister("IP");
                assert IP != null : "Come on, it can't be null...";

                if (IP.getValue() >= MEMORY.getSize()) {
                    stop();
                } else {
                    int currentIP = IP.getValue();
                    Instruction instruction = INSTRUCTIONSET.getInstruction(MEMORY.getValueAt(currentIP));
                    if (instruction == null) throw new InstructionError("Unknown", "Unknown Instruction", currentIP);
                    HISTORY.put(currentIP, instruction.getKeyword());

                    IP.setValue(currentIP + instruction.getWords());
                    try {
                        instruction.execute(
                                this,
                                // Here we check instruction.ARGUMENTS == 0 because Memory#getValuesAt always throws if the address is out of bounds
                                instruction.getArgumentsCount() == 0 ? new int[0] : MEMORY.getValuesAt(currentIP + 1, instruction.getArgumentsCount())
                        );
                    } catch (Exception err) {
                        throw new InstructionError(instruction.getKeyword(), err.getMessage(), currentIP);
                    }
                }
            }
        }
    }

    @Override
    public void stop() { isRunning = false; }

    @Override
    public boolean isPaused() {
        return this.isPaused;
    }

    @Override
    public void pause() {
        isPaused = true;
    }

    @Override
    public void resume() {
        isPaused = false;
    }

    @Override
    public void step() {
        stepping = true;
    }

}

File: plugins/RubyProcessor/Processor.rb

# Copyright (c) 2021 [hds536jhmk](https://github.com/hds536jhmk/ProcessorEmulator)
# 
# Permission is hereby granted, free of charge, to any person
# obtaining a copy of this software and associated documentation
# files (the "Software"), to deal in the Software without
# restriction, including without limitation the rights to use,
# copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following
# conditions:
# 
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
# OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
# HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
# WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR

require "PEMUJRubyPluginAPI"
require_relative "Utils.rb"

# This class is a copy-paste of PEMU::Processor (But written in Ruby):
# https://github.com/hds536jhmk/ProcessorEmulator/blob/master/src/main/java/io/github/hds/pemu/processor/Processor.java
class Processor
	protected
	attr_accessor :is_running
	attr_accessor :register_words
	attr_accessor :reserved_stack_elements
	attr_accessor :flags_words
	attr_accessor :memory
	attr_accessor :clock
	attr_accessor :instruction_set
	attr_accessor :history
	attr_accessor :registers
	attr_accessor :flags
	attr_accessor :char_pressed
	attr_accessor :key_pressed
	attr_accessor :start_timestamp
	attr_accessor :is_paused
	attr_accessor :stepping

	public
	# Implementing interface PEMU::IProcessor
	include PEMU::IProcessor

	def initialize(config)
		self.is_running = false
		self.register_words = 2
		self.reserved_stack_elements = 1
		self.flags_words = 1

		self.memory = PEMU::Memory.new(
			config.memory_size,
			PEMU::Word.getClosestWord(config.bits)
		)

		self.clock = PEMU::Clock.new config.clock_frequency

		self.instruction_set = config.instruction_set
		self.history = PEMU::InstructionHistory.new

		self.registers = PEMU::RegisterHolder.new(
			PEMU::MemoryRegister.new(getProgramAddress() , "Instruction Pointer", self.memory, 0),
			PEMU::MemoryRegister.new(self.memory.size - 1, "Stack Pointer"      , self.memory, 1)
		)

		self.flags = PEMU::FlagHolder.new(
			PEMU::MemoryFlag.new(false, "Zero Flag" , self.memory, self.register_words, 0),
			PEMU::MemoryFlag.new(false, "Carry Flag", self.memory, self.register_words, 1)
		)

		self.char_pressed = "\0"
		self.key_pressed  = PEMU::KeyEvent::VK_UNDEFINED
		self.start_timestamp = 0
		self.is_paused = false
		self.stepping  = false
	end

	def self.getDummyProcessor(config)
		return PEMU::DummyProcessor.new(
			config,
			[ PEMU::DummyMemoryRegister.new("Instruction Pointer"), PEMU::DummyMemoryRegister.new("Stack Pointer") ],
			[ PEMU::DummyMemoryFlag.new("Zero Flag"), PEMU::DummyMemoryFlag.new("Carry Flag") ]
		)
	end

	def getFlags()
		return self.flags.toArray
	end

	def getRegisters()
		return self.registers.to_a
	end

	def getFlag(shortName)
		return self.flags.getFlag shortName
	end

	def getRegister(shortName)
		return self.registers.getRegister shortName
	end

	def getMemory()
		return self.memory
	end

	def getClock()
		return self.clock
	end

	def getInstructionSet()
		return self.instruction_set
	end

	def getKeyPressed()
		return self.key_pressed
	end

	def setKeyPressed(key)
		self.key_pressed = key
	end

	def getCharPressed()
		return self.char_pressed
	end

	def setCharPressed(char)
		self.char_pressed = char
	end

	def getInfo()
        return "\tClock:\t#{ PEMU::StringUtils.getEngNotation(self.clock.getFrequency(), "Hz") }\n"\
               "\tMemory:\t#{ self.memory.getSize() }x#{ self.memory.getWord().TOTAL_BYTES } Bytes\n"\
               "\tInstructions:\t#{ self.instruction_set.getSize() }\n";
	end

	def getInstructionHistory()
		return self.history
	end

	def getTimeRunning()
		return -1 if !self.is_running
		return get_time_millis() - self.start_timestamp
	end

	def loadProgram(program)
		return "Couldn't load program because there's not enough space!" if program.length > self.memory.getSize() - getReservedWords()
		self.memory.setValuesAt(getProgramAddress(), program)
		return nil
	end

	def getProgramAddress()
		return self.register_words + self.flags_words
	end

	def getReservedWords()
		return self.register_words + self.flags_words + self.reserved_stack_elements
	end

	def isRunning()
		return self.is_running
	end

	def run()
		return if self.is_running

		self.start_timestamp = get_time_millis()
		self.is_running = true
		while self.is_running do
			if self.clock.update() && (self.stepping || !self.is_paused) then
				self.stepping = false

				ip = self.registers.getRegister("IP")

				if ip.getValue() >= self.memory.getSize() then
					stop()
				else
					currentIP = ip.getValue()
					instruction = self.instruction_set.getInstruction(self.memory.getValueAt(currentIP))
					raise PEMU::InstructionError.new("Unknown", "Unknown Instruction", currentIP) if instruction == nil
					self.history.put(currentIP.to_i, instruction.getKeyword())

					ip.setValue(currentIP + instruction.getWords())
					begin
						instruction.execute(
							self, instruction.getArgumentsCount() == 0 ? [] : self.memory.getValuesAt(currentIP + 1, instruction.getArgumentsCount())
						)
					rescue Exception => err
						raise PEMU::InstructionError.new(instruction.getKeyword(), err.to_s, currentIP)
					end
				end
			end
		end
	end

	def stop()
		self.is_running = false
	end

	def isPaused()
		return self.is_paused
	end

	def pause()
		self.is_paused = true
	end

	def resume()
		self.is_paused = false
	end

	def step()
		self.stepping = true
	end
end
